go/types/value_store.go

```diff
@@ -229,35 +242,115 @@ func (lvs *ValueStore) WriteValue(v Value) Ref {
 
 	// TODO: It _really_ feels like there should be some refactoring that allows us to only have to walk the refs of |v| once, but I'm hesitant to undertake that refactor right now.
 	hints := lvs.chunkHintsFromCache(v)
+	lvs.putGrandchildren(v, c, height, hints)
 
-	func() {
-		lvs.pendingMu.Lock()
-		defer lvs.pendingMu.Unlock()
-		lvs.pendingPuts[h] = pendingChunk{c, height, hints}
-		v.WalkRefs(func(reachable Ref) {
-			if pc, present := lvs.pendingPuts[reachable.TargetHash()]; present {
+	lvs.setHintsForReachable(v, h, true)
+	lvs.set(h, (*presentChunk)(v.Type()), false)
+	lvs.valueCache.Drop(h)
+	return r
+}
+
+func (lvs *ValueStore) putGrandchildren(v Value, c chunks.Chunk, height uint64, hints Hints) {
```

Comment:
suggestion this method be called `bufferChunk.

suggested comment:

`bufferChunk` enqueues `c` (which is the serialization of `v`) within this `ValueStores`. Buffered chunks are flushed progressively to the underlying BatchStore in a way which attempts to locate children and grandchildren sequentially together. The following invariants are retained:

1) For any given chunk currently in the buffer, only direct children of the chunk may also be presently buffered (any grandchildren will have been flushed).

2) The total data occupied by buffered chunks does not exceed `lvs.pendingPutMax'